src/ProtocolFeeController.sol

```diff
@@ -0,0 +1,117 @@
+
+/// @notice ProtocolFeeController for both Pool type
+contract ProtocolFeeController is IProtocolFeeController, Ownable {
+    using ProtocolFeeLibrary for uint24;
+
+    /// @notice throw when the pool manager saved does not match the pool manager from the pool key
+    error InvalidPoolManager();
+
+    /// @notice throw when the protocol fee split ratio is invalid i.e. greater than 100%
+    error InvliadProtocolFeeSplitRatio();
+
+    /// @notice 100% in hundredths of a bip
+    uint256 public constant ONE_HUNDRED_PERCENT_RATIO = 1e6;
+
+    /// @notice The ratio of the protocol fee in the total fee, expressed in hundredths of a bip i.e. 1e4 is 1%
+    /// @dev The default value is 33% i.e. protocol fee should be 33% of the total fee
+    uint256 public protocolFeeSplitRatio = 33 * 1e4;
+
+    /// @notice The default protocol fee to be used when a pool is initialized
+    /// warning: update this value won't affect the existing pools
+    /// @dev if not set then the protocol fee is
+    ///   1. 0 for dynamic fee pools
+    ///   2. a portion of the total fee based on 'protocolFeeSplitRatio' for static fee pools
+    mapping(PoolId => uint24) public defaultProtocolFees;
+
+    address public immutable poolManager;
+
+    /// @notice emit when the protocol fee is collected
+    event ProtocolFeeCollected(Currency indexed currency, uint256 amount);
+
+    constructor(address _clPoolManager) Ownable(msg.sender) {
+        poolManager = _clPoolManager;
+    }
+
+    /// @notice Set the ratio of the protocol fee in the total fee
+    function setProtocolFeeSplitRatio(uint256 newProtocolFeeSplitRatio) external onlyOwner {
+        if (newProtocolFeeSplitRatio > ONE_HUNDRED_PERCENT_RATIO) revert InvliadProtocolFeeSplitRatio();
+        protocolFeeSplitRatio = newProtocolFeeSplitRatio;
+    }
+
+    function setDefaultProtocolFee(PoolKey memory key, uint24 newProtocolFee) external onlyOwner {
+        if (address(key.poolManager) != poolManager) revert InvalidPoolManager();
+        /// @notice Validate the protocol fee, revert if fee is over the limit
+        if (!newProtocolFee.validate()) revert IProtocolFees.ProtocolFeeTooLarge(newProtocolFee);
+
+        defaultProtocolFees[key.toId()] = newProtocolFee;
+    }
+
+    /// @notice Get the protocol fee for a pool given the conditions of this contract
+    /// @return protocolFee The pool's protocol fee, expressed in hundredths of a bip. The upper 12 bits are for 1->0
+    function protocolFeeForPool(PoolKey memory poolKey) external view override returns (uint24 protocolFee) {
+        if (address(poolKey.poolManager) != poolManager) revert InvalidPoolManager();
+
+        // in case we've set a default protocol fee for the pool
+        uint24 defaultProtocolFee = defaultProtocolFees[poolKey.toId()];
+        if (defaultProtocolFee != 0) {
+            // already bi-directional fee so that it can be different in each direction
+            return defaultProtocolFee;
+        }
```

Comment:
1/ can you think of a scenario where we would have set `defaultProtocolFees[poolKey.toId()];` first?

2/ saying this as every pool initialization will call this method and this check means we impose this `sload` for all new pool

Reply:
For example, let's say we would like to launch our official dynamic fee hook pool, we may tend to set a more reasonable protocolFee percentage instead of 0 ?

Reply:
in this case, can we call `protocolFeeController.setProtocolFee(PoolKey memory key, uint24 newProtocolFee)` after the pool is created instead?

or maybe can do a gas check on how much extra gas would this block of code cost. All i'm saying is that this function would be called for each pool and getting everyone to pay that extra gas.. is it worth it? 

```
  uint24 defaultProtocolFee = defaultProtocolFees[poolKey.toId()];
  if (defaultProtocolFee != 0) {
     // already bi-directional fee so that it can be different in each direction
      return defaultProtocolFee;
}
```

Reply:
"in this case, can we call protocolFeeController.setProtocolFee(PoolKey memory key, uint24 newProtocolFee) after the pool is created instead?"

if this is acceptable, yeah, will just remove this mapping part

Reply:
done